### PR TITLE
[DataGrid] Change `GridColDef` methods signatures

### PR DESCRIPTION
**Pull request by cherniavskii**
Closes https://github.com/mui/mui-x/issues/10741

### Breaking changes

- The signature of `GridColDef['valueGetter']` has been changed for performance reasons:

  ```diff
  - valueGetter: ({ value, row }) => value,
  + valueGetter: (value, row, column, apiRef) => value,
  ```

  The `GridValueGetterParams` interface has been removed:

  ```diff
  - const customValueGetter = (params: GridValueGetterParams) => params.row.budget;
  + const customValueGetter: GridValueGetterFn = (value, row) => row.budget;
  ```

- The signature of `GridColDef['valueFormatter']` has been changed for performance reasons:

  ```diff
  - valueFormatter: ({ value }) => value,
  + valueFormatter: (value, row, column, apiRef) => value,
  ```

  The `GridValueFormatterParams` interface has been removed:

  ```diff
  - const gridDateFormatter = ({ value, field, id }: GridValueFormatterParams<Date>) => value.toLocaleDateString();
  + const gridDateFormatter: GridValueFormatter = (value: Date) => value.toLocaleDateString();
  ```

- The signature of `GridColDef['valueSetter']` has been changed for performance reasons:

  ```diff
  - valueSetter: (params) => {
  -   const [firstName, lastName] = params.value!.toString().split(' ');
  -   return { ...params.row, firstName, lastName };
  - }
  + valueSetter: (value, row) => {
  +   const [firstName, lastName] = value!.toString().split(' ');
  +   return { ...row, firstName, lastName };
  +}
  ```

  The `GridValueSetterParams` interface has been removed:

  ```diff
  - const setFullName = (params: GridValueSetterParams) => {
  -   const [firstName, lastName] = params.value!.toString().split(' ');
  -   return { ...params.row, firstName, lastName };
  - };
  + const setFullName: GridValueSetter<Row> = (value, row) => {
  +   const [firstName, lastName] = value!.toString().split(' ');
  +   return { ...row, firstName, lastName };
  + }
  ```

- The signature of `GridColDef['valueParser']` has been changed for performance reasons:

  ```diff
  - valueParser: (value, params: GridCellParams) => value.toLowerCase(),
  + valueParser: (value, row, column, apiRef) => value.toLowerCase(),
  ```

- The signature of `GridColDef['colSpan']` has been changed for performance reasons:

  ```diff
  - colSpan: ({ row, field, value }: GridCellParams) => (row.id === 'total' ? 2 : 1),
  + colSpan: (value, row, column, apiRef) => (row.id === 'total' ? 2 : 1),
  ```

- The signature of `GridColDef['pastedValueParser']` has been changed for performance reasons:

  ```diff
  - pastedValueParser: (value, params) => new Date(value),
  + pastedValueParser: (value, row, column, apiRef) => new Date(value),
  ```

- The signature of `GridColDef['groupingValueGetter']` has been changed for performance reasons:

  ```diff
  - groupingValueGetter: (params) => params.value.name,
  + groupingValueGetter: (value: { name: string }) => value.name,
  ```

### CODE REVIEW

**Comment by cherniavskii** [comment]
I have measured the performance impact of this PR on quick filtering by measuring the execution time of the `applyFilters` function after applying the quick filter that with no match (every column gets checked against the filtering value).
It's around 20-30% faster depending on the number of rows and columns.
Here are the detailed results:

| |  `next`  | `HEAD` | diff |
| --- | --- | --- | --- |
| **10k rows, 10 columns (no filter match)** |
| Median (ms) | 12.54 |	9.57 | -23.72% |
| Mean (ms) | 13.10 | 10.13 | -22.71% |
| σ (for 20 runs) | 1.64 |	1.76 |	
| | |
| **100k rows, 29 columns (no filter match)** |
| Median (ms) | 883.76 | 636.38 | -27.99% |
| Mean (ms) | 909.49 | 643.40 | -29.26% |
| σ (for 20 runs) | 77.06 | 22.91 |

I'm not 100% if it's worth pushing forward, because:
- most end users won't feel the difference
- breaking changes will impact almost every user

@romgrk @joserodolfofreitas What do you think?

**Comment by romgrk** [comment]
I'm still for these changes because:
 - It's 20-30% here, but they will also affect other parts of the grid wherever we use those functions.
 - Quickfilters aren't usable for larger datasets, 800+ms is horribly long. And I imagine you run those numbers of your (fast) computer.
 - The complexity of the change for users seems quite low.

**Review by romgrk — APPROVED** [review]

File: packages/grid/x-data-grid-premium/src/models/gridGroupingValueGetter.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import {
+  // GridRowId,
```
Comment (MBilalShafi) [inline packages/grid/x-data-grid-premium/src/models/gridGroupingValueGetter.ts:2]: Do we still need these commented imports?

File: packages/grid/x-data-grid/src/hooks/features/rows/useGridParamsApi.ts [hunk 2]
```diff
@@ -26,10 +23,8 @@ export class MissingRowIdError extends Error {}
  */
 export function useGridParamsApi(
   apiRef: React.MutableRefObject<GridPrivateApiCommunity>,
-  props: Pick<DataGridProcessedProps, 'getRowId'>,
+  // props: Pick<DataGridProcessedProps, 'getRowId'>,
```
Comment (MBilalShafi) [inline packages/grid/x-data-grid/src/hooks/features/rows/useGridParamsApi.ts:26]:
```suggestion
```

**Review by MBilalShafi — APPROVED** [review]
Nice improvement 🚀
I think it makes sense to add these changes to the migration guide too

**Comment by quintonlucyk** [comment]
Is this documentation up to date? I'm encountering `Module '"@mui/x-data-grid-premium"' has no exported member 'GridColDef'.`
image

**Comment by romgrk** [comment]
Open a new issue for assistance.

**Pull request merged**